File: Source/ILTranslucentView.m
//
//  ILTranslucentView.m
//  ILTranslucentViewProject
//
//  Created by Ivo Leko on 10/11/13.
//  Copyright (c) 2013 Ivo Leko. All rights reserved.
//

#import "ILTranslucentView.h"
#define SYSTEM_VERSION_GREATER_THAN_OR_EQUAL_TO(v)  ([[[UIDevice currentDevice] systemVersion] compare:v options:NSNumericSearch] != NSOrderedAscending)

@interface ILTranslucentView () {
    UIView *nonexistentSubview; //this is first child view of ILTranslucentView. It is not available trough "Managing the View Hierarchy" methods.
    UIView *toolbarContainerClipView; //view which contain UIToolbar as child subview
    UIToolbar *toolBarBG; //this is empty toolbar which we use to produce blur (translucent) effect
    UIView *overlayBackgroundView; //view over toolbar that is responsive to backgroundColor property
    BOOL initComplete;
}

@property (nonatomic, copy) UIColor *ilColorBG; //backGround color
@property (nonatomic, copy) UIColor *ilDefaultColorBG; //default Apple's color of UIToolbar

@end

@implementation ILTranslucentView
@synthesize translucentAlpha = _translucentAlpha;


#pragma mark - Initalization
- (id)initWithFrame:(CGRect)frame //code
{
    self = [super initWithFrame:frame];
    if (self) {
        [self createUI];
    }
    return self;
}

- (id) initWithCoder:(NSCoder *)aDecoder { //XIB
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self createUI];
    }
    return self;
}


- (void) createUI {
    
    if (SYSTEM_VERSION_GREATER_THAN_OR_EQUAL_TO(@"7.0")) {
        _ilColorBG = self.backgroundColor; //get background color of view (can be set trough interface builder before)
        self.ilDefaultColorBG=toolBarBG.barTintColor; //Apple's default background color
        
        _translucent=YES;
        _translucentAlpha=1;
        
        // creating nonexistentSubview
        nonexistentSubview = [[UIView alloc] initWithFrame:self.bounds];
        nonexistentSubview.backgroundColor=[UIColor clearColor];
        nonexistentSubview.clipsToBounds=YES;
        nonexistentSubview.autoresizingMask = UIViewAutoresizingFlexibleBottomMargin | UIViewAutoresizingFlexibleLeftMargin | UIViewAutoresizingFlexibleRightMargin | UIViewAutoresizingFlexibleTopMargin;
        [self insertSubview:nonexistentSubview atIndex:0];
        
        //creating toolbarContainerClipView
        toolbarContainerClipView = [[UIView alloc] initWithFrame:self.bounds];
        toolbarContainerClipView.backgroundColor=[UIColor clearColor];
        toolbarContainerClipView.clipsToBounds=YES;
        toolbarContainerClipView.autoresizingMask = UIViewAutoresizingFlexibleBottomMargin | UIViewAutoresizingFlexibleLeftMargin | UIViewAutoresizingFlexibleRightMargin | UIViewAutoresizingFlexibleTopMargin;
        [nonexistentSubview addSubview:toolbarContainerClipView];
        
        //creating toolBarBG
        //we must clip 1px line on the top of toolbar
        CGRect rect= self.bounds;
        rect.origin.y-=1;
        rect.size.height+=1;
        toolBarBG =[[UIToolbar alloc] initWithFrame:rect];
        toolBarBG.frame=rect;
        toolBarBG.autoresizingMask= UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
        [toolbarContainerClipView addSubview:toolBarBG];
        
    
        
        
        //view above toolbar, great for changing blur color effect
        overlayBackgroundView = [[UIView alloc] initWithFrame:self.bounds];
        overlayBackgroundView.backgroundColor = self.backgroundColor;
        overlayBackgroundView.autoresizingMask= UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
        [toolbarContainerClipView addSubview:overlayBackgroundView];
        
        
        [self setBackgroundColor:[UIColor clearColor]]; //view must be transparent :)
        initComplete=YES;
    }
    
}

#pragma mark - Configuring a View’s Visual Appearance



- (BOOL) isItClearColor: (UIColor *) color {
    CGFloat red = 0.0, green = 0.0, blue = 0.0, alpha =0.0;
    [color getRed:&red green:&green blue:&blue alpha:&alpha];
    
    if (red!=0 || green != 0 || blue != 0 || alpha != 0) {
        return NO;
    }
    else {
        return YES;
    }
}

- (void) setFrame:(CGRect)frame {
    
    
    //     - Setting frame of view -
    // UIToolbar's frame is not great at animating. It produces lot of glitches.
    // Because of that, we never actually reduce size of toolbar"
    
    if (SYSTEM_VERSION_GREATER_THAN_OR_EQUAL_TO(@"7.0")) {
        CGRect rect = frame;
        rect.origin.x=0;
        rect.origin.y=0;
        
        if (toolbarContainerClipView.frame.size.width>rect.size.width) {
            rect.size.width=toolbarContainerClipView.frame.size.width;
        }
        if (toolbarContainerClipView.frame.size.height>rect.size.height) {
            rect.size.height=toolbarContainerClipView.frame.size.height;
        }
        
        toolbarContainerClipView.frame=rect;
        
        [super setFrame:frame];
        [nonexistentSubview setFrame:self.bounds];
    }
    else
        [super setFrame:frame];
}

- (void) setBounds:(CGRect)bounds {
    
    //     - Setting bounds of view -
    // UIToolbar's bounds is not great at animating. It produces lot of glitches.
    // Because of that, we never actually reduce size of toolbar"
    
    
    if (SYSTEM_VERSION_GREATER_THAN_OR_EQUAL_TO(@"7.0")) {
        CGRect rect = bounds;
        rect.origin.x=0;
        rect.origin.y=0;
        
        if (toolbarContainerClipView.bounds.size.width>rect.size.width) {
            rect.size.width=toolbarContainerClipView.bounds.size.width;
        }
        if (toolbarContainerClipView.bounds.size.height>rect.size.height) {
            rect.size.height=toolbarContainerClipView.bounds.size.height;
        }
        
        toolbarContainerClipView.bounds=rect;
        [super setBounds:bounds];
        [nonexistentSubview setFrame:self.bounds];
    }
    else
        [super setBounds:bounds];
    
}


- (void) setTranslucentStyle:(UIBarStyle)translucentStyle {
    toolBarBG.barStyle=translucentStyle;
}
- (UIBarStyle) translucentStyle {
    return toolBarBG.barStyle;
}


- (void) setTranslucentTintColor:(UIColor *)translucentTintColor {
   
    _translucentTintColor = translucentTintColor;
    
    //tint color of toolbar
    if ([self isItClearColor:translucentTintColor])
        [toolBarBG setBarTintColor:self.ilDefaultColorBG];
    else
        [toolBarBG setBarTintColor:translucentTintColor];
}

- (void) setBackgroundColor:(UIColor *)backgroundColor {
    
    //changing backgroundColor of view actually change tintColor of toolbar
    if (initComplete) {
        
        self.ilColorBG=backgroundColor;
        if (_translucent) {
            overlayBackgroundView.backgroundColor = backgroundColor;
            [super setBackgroundColor:[UIColor clearColor]];
        }
        else {
            [super setBackgroundColor:self.ilColorBG];
        }
        
    }
    else
        [super setBackgroundColor:backgroundColor];
}

- (void) setTranslucent:(BOOL)translucent {
    
    //enabling and disabling blur effect
    _translucent=translucent;
    
    toolBarBG.translucent=translucent;
    if (translucent) {
        toolBarBG.hidden=NO;
        [toolBarBG setBarTintColor:self.ilColorBG];
        [self setBackgroundColor:[UIColor clearColor]];
    }
    else {
        toolBarBG.hidden=YES;
        [self setBackgroundColor:self.ilColorBG];
    }
}

- (void) setTranslucentAlpha:(CGFloat)translucentAlpha {
    //changing alpha of toolbar
    
    if (translucentAlpha>1)
        _translucentAlpha=1;
    else if (translucentAlpha<0)
        _translucentAlpha=0;
    else
        _translucentAlpha=translucentAlpha;
    
    toolBarBG.alpha=translucentAlpha;

}


#pragma mark - Managing the View Hierarchy

- (NSArray *) subviews {

    // must exclude nonexistentSubview
    
    if (initComplete) {
        NSMutableArray *array = [NSMutableArray arrayWithArray:[super subviews]];
        [array removeObject:nonexistentSubview];
        return (NSArray *)array;
    }
    else {
        return [super subviews];
    }
    
}
- (void) sendSubviewToBack:(UIView *)view {
    
    // must exclude nonexistentSubview

    if (initComplete) {
        [self insertSubview:view aboveSubview:toolbarContainerClipView];
        return;
    }
    else
        [super sendSubviewToBack:view];
}
- (void) insertSubview:(UIView *)view atIndex:(NSInteger)index {
    
    // must exclude nonexistentSubview
    
    if (initComplete) {
        [super insertSubview:view atIndex:(index+1)];
    }
    else
        [super insertSubview:view atIndex:index];

}
- (void) exchangeSubviewAtIndex:(NSInteger)index1 withSubviewAtIndex:(NSInteger)index2 {
    
    // must exclude nonexistentSubview
    
    if (initComplete)
        [super exchangeSubviewAtIndex:(index1+1) withSubviewAtIndex:(index2+1)];
    else
        [super exchangeSubviewAtIndex:(index1) withSubviewAtIndex:(index2)];
}


/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect
{
    // Drawing code
}
*/

@end

File: Source/ILTranslucentView.swift
//
//  ILTranslucentView.swift
//  ILTranslucentView
//
//  Created by Tomasz Szulc on 25/07/14.
//  http://github.com/tomkowz
//

/// Rewritten from https://github.com/ivoleko/ILTranslucentView

import UIKit

@IBDesignable
open class ILTranslucentView: UIView {

    fileprivate var _translucent = true
    @IBInspectable open var translucent : Bool {
        set {
            _translucent = newValue
            if self.toolbarBG == nil {
                return
            }

            self.toolbarBG!.isTranslucent = newValue

            if newValue {
                self.toolbarBG!.isHidden = false
                self.toolbarBG!.barTintColor = self.ilColorBG
                self.backgroundColor = UIColor.clear
            } else {
                self.toolbarBG!.isHidden = true
                self.backgroundColor = self.ilColorBG
            }
        }
        get {
            return _translucent
        }
    }

    fileprivate var _translucentAlpha : CGFloat = 1.0
    @IBInspectable open var translucentAlpha : CGFloat {
        set {
            if newValue > 1 {
                _translucentAlpha = 1
            } else if (newValue < 0) {
                _translucentAlpha = 0
            } else {
                _translucentAlpha = newValue
            }

            if self.toolbarBG != nil {
                self.toolbarBG!.alpha = _translucentAlpha
            }
        }
        get {
            return _translucentAlpha
        }
    }

    @IBInspectable open var translucentStyle : UIBarStyle {
        set {
            if self.toolbarBG != nil {
                self.toolbarBG!.barStyle = newValue
            }
        }
        get {
            if self.toolbarBG != nil {
                return self.toolbarBG!.barStyle
            } else {
                return UIBarStyle.default
            }
        }
    }

    fileprivate var _translucentTintColor = UIColor.clear
    @IBInspectable open var translucentTintColor : UIColor {
        set {
            _translucentTintColor = newValue
            if (self.isItClearColor(newValue)) {
                self.toolbarBG!.barTintColor = self.ilDefaultColorBG
            } else {
                self.toolbarBG!.barTintColor = self.translucentTintColor
            }
        }
        get {
            return _translucentTintColor
        }
    }

    fileprivate var ilColorBG : UIColor?
    fileprivate var ilDefaultColorBG : UIColor?

    fileprivate var toolbarBG : UIToolbar?
    fileprivate var nonExistentSubview : UIView?
    fileprivate var toolbarContainerClipView : UIView?
    fileprivate var overlayBackgroundView : UIView?
    fileprivate var initComplete = false

    override init(frame: CGRect) {
        super.init(frame: frame)
        self.createUI()
    }

    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        self.createUI()
    }
}

extension ILTranslucentView {
    fileprivate func createUI() {
        self.ilColorBG = self.backgroundColor

        self.translucent = true
        self.translucentAlpha = 1

        let _nonExistentSubview = UIView(frame: self.bounds)
        _nonExistentSubview.backgroundColor = UIColor.clear
        _nonExistentSubview.clipsToBounds = true
        _nonExistentSubview.autoresizingMask = [UIViewAutoresizing.flexibleBottomMargin, UIViewAutoresizing.flexibleLeftMargin, UIViewAutoresizing.flexibleRightMargin, UIViewAutoresizing.flexibleTopMargin]
        self.nonExistentSubview = _nonExistentSubview
        self.insertSubview(self.nonExistentSubview!, at: 0)

        let _toolbarContainerClipView = UIView(frame: self.bounds)
        _toolbarContainerClipView.backgroundColor = UIColor.clear
        _toolbarContainerClipView.clipsToBounds = true
        _toolbarContainerClipView.autoresizingMask = [UIViewAutoresizing.flexibleBottomMargin, UIViewAutoresizing.flexibleLeftMargin, UIViewAutoresizing.flexibleRightMargin, UIViewAutoresizing.flexibleTopMargin]
        self.toolbarContainerClipView = _toolbarContainerClipView
        self.nonExistentSubview!.addSubview(self.toolbarContainerClipView!)

        var rect = self.bounds
        rect.origin.y -= 1
        rect.size.height += 1

        let _toolbarBG = UIToolbar(frame: rect)
        _toolbarBG.autoresizingMask = [UIViewAutoresizing.flexibleWidth, UIViewAutoresizing.flexibleHeight]
        self.toolbarBG = _toolbarBG

        self.toolbarContainerClipView!.addSubview(self.toolbarBG!)
        self.ilDefaultColorBG = self.toolbarBG!.barTintColor

        let _overlayBackgroundView = UIView(frame: self.bounds)
        _overlayBackgroundView.backgroundColor = self.backgroundColor
        _overlayBackgroundView.autoresizingMask = [UIViewAutoresizing.flexibleWidth, UIViewAutoresizing.flexibleHeight]
        self.overlayBackgroundView = _overlayBackgroundView
        self.toolbarContainerClipView!.addSubview(self.overlayBackgroundView!)

        self.backgroundColor = UIColor.clear
        self.initComplete = true
    }

    fileprivate func isItClearColor(_ color: UIColor) -> Bool {
        var red : CGFloat = 0.0
        var green : CGFloat = 0.0
        var blue : CGFloat = 0.0
        var alpha : CGFloat = 0.0
        color.getRed(&red, green: &green, blue: &blue, alpha: &alpha)
        return red == 0.0 && green == 0.0 && blue == 0.0 && alpha == 0.0
    }

    @objc open override var frame : CGRect {
        set {
            if self.toolbarContainerClipView == nil {
                super.frame = newValue
                return
            }

            var rect = newValue
            rect.origin = CGPoint.zero

            let width = self.toolbarContainerClipView!.frame.width
            if width > rect.width {
                rect.size.width = width
            }

            let height = self.toolbarContainerClipView!.frame.height
            if height > rect.height {
                rect.size.height = height
            }

            self.toolbarContainerClipView!.frame = rect

            super.frame = newValue
            self.nonExistentSubview!.frame = self.bounds
        }
        get {
            return super.frame
        }
    }

    @objc override open var bounds : CGRect {
        set {
            var rect = newValue
            rect.origin = CGPoint.zero

            let width = self.toolbarContainerClipView!.bounds.width
            if width > rect.width {
                rect.size.width = width
            }

            let height = self.toolbarContainerClipView!.bounds.height
            if height > rect.height {
                rect.size.height = height
            }

            self.toolbarContainerClipView!.bounds = rect
            super.bounds = newValue
            self.nonExistentSubview!.frame = self.bounds
        }
        get {
            return super.bounds
        }
    }

    @objc override open var backgroundColor : UIColor! {
        set {
            if self.initComplete {
                self.ilColorBG = newValue
                if (self.translucent) {
                    self.overlayBackgroundView!.backgroundColor = newValue
                    super.backgroundColor = UIColor.clear
                }
            } else {
                super.backgroundColor = self.ilColorBG
            }
        }
        get {
            return super.backgroundColor
        }
    }

    @objc override open var subviews : [UIView] {
        if self.initComplete {
            var array = super.subviews as [UIView]

            var index = 0
            for view in array {
                if view == self.nonExistentSubview {
                    break
                }
                index += 1
            }

            if index < array.count {
                array.remove(at: index)
            }

            return array
        } else {
            return super.subviews
        }
    }

    override open func sendSubview(toBack view: UIView)  {
        if self.initComplete {
            self.insertSubview(view, aboveSubview: self.toolbarContainerClipView!)
        } else {
            super.sendSubview(toBack: view)
        }
    }

    override open func insertSubview(_ view: UIView, at index: Int) {
        if self.initComplete {
            super.insertSubview(view, at: index + 1)
        } else {
            super.insertSubview(view, at: index)
        }
    }

    override open func exchangeSubview(at index1: Int, withSubviewAt index2: Int)  {
        if self.initComplete {
            super.exchangeSubview(at: (index1 + 1), withSubviewAt: (index2 + 1))
        } else {
            super.exchangeSubview(at: index1, withSubviewAt: index2)
        }
    }
}
